dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/service/DefaultTrackerImportAccessManager.java

```diff
@@ -0,0 +1,254 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.hisp.dhis.category.CategoryOption;
+import org.hisp.dhis.category.CategoryOptionCombo;
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+import org.hisp.dhis.organisationunit.OrganisationUnitService;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.program.ProgramStage;
+import org.hisp.dhis.program.ProgramStageInstance;
+import org.hisp.dhis.security.acl.AclService;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.trackedentity.TrackerOwnershipManager;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.report.ValidationErrorReporter;
+import org.hisp.dhis.tracker.validation.hooks.Constants;
+import org.hisp.dhis.user.User;
+import org.springframework.stereotype.Component;
+
+import java.util.Objects;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.hisp.dhis.tracker.report.ValidationErrorReporter.newReport;
+import static org.hisp.dhis.tracker.validation.hooks.Constants.PROGRAM_CANT_BE_NULL;
+import static org.hisp.dhis.tracker.validation.hooks.Constants.PROGRAM_INSTANCE_CANT_BE_NULL;
+import static org.hisp.dhis.tracker.validation.hooks.Constants.PROGRAM_STAGE_CANT_BE_NULL;
+import static org.hisp.dhis.tracker.validation.hooks.Constants.PROGRAM_STAGE_INSTANCE_CANT_BE_NULL;
+import static org.hisp.dhis.tracker.validation.hooks.Constants.USER_CANT_BE_NULL;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Component
+public class DefaultTrackerImportAccessManager
+    implements TrackerImportAccessManager
+{
+    private final AclService aclService;
+
+    private final TrackerOwnershipManager ownershipAccessManager;
+
+    private final OrganisationUnitService organisationUnitService;
+
+    public DefaultTrackerImportAccessManager( AclService aclService, TrackerOwnershipManager ownershipAccessManager,
+        OrganisationUnitService organisationUnitService )
+    {
+        checkNotNull( aclService );
+        checkNotNull( ownershipAccessManager );
+        checkNotNull( organisationUnitService );
+
+        this.aclService = aclService;
+        this.ownershipAccessManager = ownershipAccessManager;
+        this.organisationUnitService = organisationUnitService;
+    }
+
+    @Override
+    public void checkReadEnrollmentAccess( ValidationErrorReporter reporter, User user,
+        ProgramInstance programInstance )
+    {
+        Objects.requireNonNull( user, USER_CANT_BE_NULL );
+        Objects.requireNonNull( programInstance, PROGRAM_INSTANCE_CANT_BE_NULL );
+        Objects.requireNonNull( programInstance.getProgram(), PROGRAM_CANT_BE_NULL );
+
+        checkProgramReadAccess( reporter, user, programInstance.getProgram() );
```

Comment:
Maybe we should ensure the preheater cant return data we dont have read access to?

Reply:
This goes for all the "checkRead" code.

Reply:
I personally think that introducing a security layer into the preheater is going to mix up things too much now. There is not necessary (a hard requirement) that all objects being checked is coming from the preheater, so we would then have to introduce extra logic and code to make sure we check correctly what type of objects we check, since they look the same this could be difficult etc. I think that maybe a bigger overhaul of the security model is something to plan for a separate task.